engine/jobmaster/dm/ddl_coordinator.go

```diff
@@ -0,0 +1,841 @@
+	if err != nil {
+		return nil, err
+	}
+	c.shardGroups[targetTable] = newGroup
+	return newGroup, nil
+}
+
+func (c *DDLCoordinator) removeShardGroup(ctx context.Context, targetTable metadata.TargetTable) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	if g, ok := c.shardGroups[targetTable]; ok {
+		if err := g.clear(ctx); err != nil {
+			c.logger.Error("clear shard group failed", zap.Error(err))
+		}
+	}
+	delete(c.shardGroups, targetTable)
+}
+
+type shardGroup struct {
+	mu                  sync.RWMutex
+	normalTables        map[metadata.SourceTable]string
+	conflictTables      map[metadata.SourceTable]string
+	tableAgent          TableAgent
+	droppedColumnsStore *metadata.DroppedColumnsStore
+	id                  frameModel.MasterID
+	cfg                 *config.JobCfg
+	deleted             bool
+}
+
+func newShardGroup(ctx context.Context, id frameModel.MasterID, cfg *config.JobCfg, targetTable metadata.TargetTable, sourceTables map[metadata.SourceTable]struct{}, kvClient metaModel.KVClient, tableAgent TableAgent) (*shardGroup, error) {
+	g := &shardGroup{
+		tableAgent:          tableAgent,
+		normalTables:        make(map[metadata.SourceTable]string),
+		conflictTables:      make(map[metadata.SourceTable]string),
+		droppedColumnsStore: metadata.NewDroppedColumnsStore(kvClient, targetTable),
+		id:                  id,
+		cfg:                 cfg,
+	}
+	for sourceTable := range sourceTables {
+		stmt, err := g.tableAgent.FetchTableStmt(ctx, id, cfg, sourceTable)
+		// NOTE: There are cases where the source table exists but the table info does not,
+		// such as a table is created upstream when the coordinator starts, but the dm-worker is not yet synchronized
+		// we skip these tables in handleDDL.
+		// TODO: better error handling
+		if err != nil && !strings.HasPrefix(err.Error(), "table info not found") {
+			return nil, err
+		}
+		g.normalTables[sourceTable] = stmt
+	}
+	return g, nil
+}
+
+func (g *shardGroup) handle(ctx context.Context, item *metadata.DDLItem) ([]string, optimism.ConflictStage, bool, error) {
+	// each group only processes one item at a time
+	g.mu.Lock()
+	defer g.mu.Unlock()
+
+	if g.deleted {
+		return nil, optimism.ConflictError, false, errors.Errorf("shard group for target table %v is deleted", item.TargetTable)
+	}
+
+	// nolint:errcheck
+	g.gcDroppedColumns(ctx)
+
+	var (
+		ddls          []string
+		needDeleted   bool
+		err           error
+		conflictStage optimism.ConflictStage = optimism.ConflictNone
+	)
+
+	switch item.Type {
+	case metadata.CreateTable:
+		g.handleCreateTable(item)
+		ddls = append(ddls, item.DDLs...)
+	case metadata.DropTable:
+		needDeleted = g.handleDropTable(ctx, item)
+		ddls = append(ddls, item.DDLs...)
+	case metadata.OtherDDL:
+		ddls, conflictStage, err = g.handleDDLs(ctx, item)
+	default:
+		return nil, optimism.ConflictError, false, errors.Errorf("unknown ddl type %v", item.Type)
+	}
+
+	return ddls, conflictStage, needDeleted || g.isResolved(ctx), err
```

Comment:
what's the definition of resolved? without knowing it developers can't understand why it can be combined with needDeleted?